pkg/receive/handler.go

```diff
@@ -1310,17 +1357,21 @@ func (p *peerGroup) getConnection(ctx context.Context, addr string) (WriteableSt
 	defer p.m.Unlock()
 	// Make sure that another caller hasn't created the connection since obtaining the write lock.
 	c, ok = p.connections[addr]
-	if ok {
+	if ok && c.isReady() {
 		return c, nil
 	}
+	level.Debug(p.logger).Log("msg", "dialing peer", "addr", addr)
 	conn, err := p.dialer(ctx, addr, p.dialOpts...)
```

Comment:
I searched in the repo. This is the only place where `p.dialer` is called. That means the handler never re-establishs a connection to a DNS. Before your change, the write errors eventually die down. That means the gRPC framework eventually gets the correct IP and re-establishs the connection under the hood, but the delay is quite long and causes many write errors. 
I'm concerned that there might be a reason that the gRPC call site (receive handler) doesn't explicitly dial again. Should we ask the community?

Reply:
I suspect the default gRPC dial options in Thanos is not ideal. Note that `Unavailable`(503) is retriable. 
<img width="1390" alt="image" src="https://github.com/databricks/thanos/assets/129140017/2dbd9a47-a8f5-43e2-ab70-3e06108db84d">

Reply:
seems it is way long and probably on purpose to avoid busy reconnection during normal period:

<img width="851" alt="Screenshot 2024-05-13 at 8 34 52 PM" src="https://github.com/databricks/thanos/assets/96499497/7f81d01e-a467-49a8-933c-49b0038febfe">

Reply:
actually the default idle timeout to 5 Minutes is the root cause, i was able to pass the unit tests with 1 second idle timeout.